src/spartan/polynomial.rs

```diff
@@ -145,19 +185,83 @@ impl<Scalar: PrimeField> SparsePolynomial<Scalar> {
   pub fn evaluate(&self, r: &[Scalar]) -> Scalar {
     assert_eq!(self.num_vars, r.len());
 
-    let get_bits = |num: usize, num_bits: usize| -> Vec<bool> {
-      (0..num_bits)
-        .into_par_iter()
-        .map(|shift_amount| ((num & (1 << (num_bits - shift_amount - 1))) > 0))
-        .collect::<Vec<bool>>()
-    };
-
     (0..self.Z.len())
       .into_par_iter()
       .map(|i| {
-        let bits = get_bits(self.Z[i].0, r.len());
+        let bits = (self.Z[i].0).get_bits(r.len());
         SparsePolynomial::compute_chi(&bits, r) * self.Z[i].1
       })
       .reduce(|| Scalar::ZERO, |x, y| x + y)
   }
 }
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+
+  #[derive(PrimeField)]
+  #[PrimeFieldModulus = "52435875175126190479447740508185965837690552500527637822603658699938581184513"]
+  #[PrimeFieldGenerator = "7"]
+  #[PrimeFieldReprEndianness = "little"]
+  struct Fp([u64; 4]);
+
+  #[test]
+  fn test_eq_polynomial() {
+    let ZERO = Fp::from(0);
+    let ONE = Fp::from(1);
+
+    let eq_poly = EqPolynomial::<Fp>::new(vec![ONE, ZERO, ONE]);
+    let y = eq_poly.evaluate(vec![ONE, ONE, ONE].as_slice());
+    assert_eq!(y, ZERO);
+
+    let y = eq_poly.evaluate(vec![ONE, ZERO, ONE].as_slice());
+    assert_eq!(y, ONE);
+
+    let eval_list = eq_poly.evals();
+    for i in 0..(2_usize).pow(3) {
+      if i == 5 {
+        assert_eq!(eval_list[i], ONE);
+      } else {
+        assert_eq!(eval_list[i], ZERO);
+      }
+    }
+  }
+
+  #[test]
+  fn test_multilinear_polynomial() {
+    // Let the polynomial has 3 variables, p(x_1, x_2, x_3) = (x_1 + x_2) * x_3
+    // Evaluations of the polynomial at boolean cube are [0, 0, 0, 1, 0, 1, 0, 2].
+
+    let ZERO = Fp::from(0);
+    let ONE = Fp::from(1);
```

Comment:
These are already associated constants in `Field` trait.

Reply:
Thank you for the tip!